idaes/apps/grid_integration/bidder.py

```diff
@@ -0,0 +1,774 @@
+# Xian Gao
+# Dowling Lab
+# xgao1@nd.edu
+
+import pandas as pd
+import pyomo.environ as pyo
+import os
+from collections import deque, OrderedDict
+from itertools import combinations
+
+def get_data_given(df, bus=None, date=None, hour=None, generator=None, fuel_type=None):
+
+    '''
+    This function gets the data out of a pandas dataframe given one or more
+    options, e.g. time.
+
+    Arguments:
+        df: the dataframe we are interested in
+        bus: the bus ID we want [int]/ the bus name we want [str]
+        date: the date we want [str]
+        hour: the hour we want [int]
+        generator: the generator ID [str]
+        fuel_type: generator fuel, e.g. Coal [str]
+    Returns:
+        df: a dataframe that has the information we specified.
+    '''
+
```

Comment:
Consider performing checks here on if the data frame contains the necessary fields.